### ISSUE TITLE
Support for shrinking filesystem

### ISSUE THREAD

**Issue by thememika** [koverstreet/bcachefs#781]
Shrinking the filesystem seems to be one of very important, core features. For some reason, it's still not even in the Roadmap: https://bcachefs.org/Roadmap
What is the status of this feature, and is it planned at all?
Thanks!

**Comment by Nihon-Ryori** [comment]
See the Bcachefs manual (dont have a revision number or date ...):
3.6 Device management
3.6.1 Filesystem resize
A filesystem can be resized on a particular device with the bcachefs device
resize subcommand. Currently only growing is supported, not shrinking.
* https://bcachefs.org/bcachefs-principles-of-operation.pdf

**Comment by Syntrait** [comment]
I think adding a shrinking feature would be very useful, since people that dual-boot often use GParted to move free space between their Linux and Windows partitions.

**Comment by Syntrait** [comment]
Or if you have unlimited internet, taking note of and then deleting all the games, documents, files that you can find and redownload again, and then backing up what's left, like browser data, personal documents or config files. It's possible to just archive it, encrypt it, and then upload to a cloud, if you don't have another drive. The rest is basically what @nikp123 said.

**Comment by nikp123** [comment]
Yeah didn't say that this is a solution but a workaround if your bcachefs partition has multiple drives. Bcachefs very much needs a shrink feature, this is half the reason why i never committed to f2fs for example (it still doesn't do shrinking IIRC)

**Comment by hgkamath** [comment]
- Not being able to shrink can be a cause of hesitation to install for permanent long term use on a workstation. 
- On server systems, once the planning phase is over and the decision is made to go with bcachefs, it makes sense to dispense with the need to ever consider modifying the disk layout, or trying out other filesystems, until a long distant time n the future that necessitates a server-rebuild. Server deployments are usually not intended for experimentation.
- For personal workstations on the other hand, a machine owner may want freedom in being able to run dual boot OS-es, move partitions around, make space to try new partitions for different purposes. 
- The same freedom, that is to shrink the other partitions like ntfs, fat32, ext4, btrfs, the use of which made  it possible to make space for a bcachefs partition, isn't reciprocated by bcachefs. Presently bcachefs allows for one-way growth. A strategy would be to start with as small a partition as possible to fit the OS and cover its daily excesses like updates and only grow as needed. In 2 disk machines, one ssd and one hdd, one would want to ideally configure it to use a bcachefs cache partition on the ssd.  Being stingy on space might limit what bcachefs could do. For a user to find what works best, the user will need to experiment, but to experiment one requires flexibility. 
- This keeps me from using xfs for the same reason, and so I end up using btrfs for primary purposes, which does shrink. I occasionally try out bcachefs on temporary basis and delete the disk-partition/VM-disk-partition after trying it out for a short term.

**Comment by matthiasgoergens** [comment]
We can probably support shrinking as a three-step process, thanks to already supporting multi-device operations:

1. split an existing partition in twain
2. evacuate one of the two partitions and once emptied, delete it.
3. delete the evacuated partition

Only the first step is new.  I think in principle it would only involve operations on metadata, and not any moving of non-meta data.  (Apart from moving perhaps O(1) pieces of data around that happen to straddle the border between the two new partitions.)

**Comment by Nihon-Ryori** [comment]
In my opinion, the following is a possible solution to realize shrinking of partitions for bcachefs:

1. divide the partition to be downsized into a partition with the desired target size plus a partition with the corresponding remaining size.
2. move the data of the partition to be deleted later to the partition to be retained.
3. delete the emptied partition.

Once this has been implemented in bcachefs, it can also be integrated into GParted.

**Comment by matthiasgoergens** [comment]
> In my opinion, the following is a possible solution to realize shrinking of partitions for bcachefs: [...]

Is that essentially what I said, or am I missing a subtlety?

Btw you can use the same idea to move a partition to a new place, even if old and new overlap. However, you might also want to implement being able to merge adjacent partitions for completeness. (Not sure if the logic would be the same.)

Of course, this is all easy for us to speculate about. Implementation is another question.

**Comment by Nihon-Ryori** [comment]
> Is that essentially what I said, or am I missing a subtlety?

It may be the same for native speakers, but my translator had problems translating it for me. So it seems that everything is ok in both versions. Let's keep looking forward.

**Comment by iav** [comment]
> 1. split an existing partition in twain

This is completely pointless, as it requires a "split partition" function, which doesn't exist, and most likely never will, since who and why might it even be necessary? 
Can you name a file system that has this feature?

**Comment by Silverdev2482** [comment]
> Or if you have unlimited internet, taking note of and then deleting all the games, documents, files that you can find and redownload again, and then backing up what's left, like browser data, personal documents or config files. It's possible to just archive it, encrypt it, and then upload to a cloud, if you don't have another drive. The rest is basically what [@nikp123](https://github.com/nikp123) said.

I think its actually half useful, way easier than reinstall if you have the drives and storage space to spare.

**Comment by Franck666666** [comment]
the Ai told Kent never did read or view this here , but he more likes anyway having Trouble with Linus , Ai told so when asked LoL

**Comment by cyans-nya** [comment]
Supposely shrinking is just to reallocate data to be removed to free space in new part.
And also move pointers?

**Comment by koverstreet** [comment]
If anyone wants to work on it shrink is pretty easy to add via bch2_move_data_phys(), I've just had higher priority things that need doing :)

**Comment by Komzpa** [comment]
Current related work I verified while triaging this:

- Kernel WIP: https://github.com/koverstreet/bcachefs/pull/1073 (`bcachefs: implement online filesystem shrinking`). It is open but currently `mergeStateStatus: DIRTY`; Kent reviewed it as the right reconcile-based direction, with remaining work around cached-data handling in the shrink region, progress/wakeup instead of blind polling, journal/resume cases, cleanup of commented/TODO/unrelated changes, and more ktest cases.
- Tools bridge: https://github.com/koverstreet/bcachefs-tools/pull/650 (`device: let online resize delegate shrink to kernel`). It is open, `mergeStateStatus: CLEAN`, and the full current CI matrix is green. It does not implement shrink by itself; it removes the userspace online-shrink precheck so `bcachefs device resize` can delegate smaller target sizes to a kernel that supports shrink, while offline shrink still stays rejected in tools.

So the issue is still open at the kernel implementation level, but the tools-side command path is ready for the kernel feature once koverstreet/bcachefs#1073 or a successor lands.

**Comment by jullanggit** [comment]
Thanks @Komzpa, some small updates:
The kernel-side implementation progressed quite a bit since Kent's review, it now passes the entire [shrink-specific test suite](https://github.com/jullanggit/ktest/tree/shrink), although still somewhat flakily. I'm currently actively working on getting the implementation to be more solid. Right now I'd say it is ready for some experimentation, but I would not recommend running it on any important system yet.

In [my bcachefs-tools fork](https://github.com/jullanggit/bcachefs-tools) I have also removed the online precheck (altough without adding as much context as you did), and additionally added functionality easily cancel an ongoing resize. I'm also actively working on pulling in the shrinking changes into the userspace implementation.

Help on any of this is of course welcome :)

**Comment by Komzpa** [comment]
Tools-targeted WIP artifact is updated at https://github.com/koverstreet/bcachefs-tools/pull/742.

Current status:

- `koverstreet/bcachefs-tools#742` ports the online shrink implementation from `koverstreet/bcachefs#1073` into `bcachefs-tools/testing`.
- `koverstreet/bcachefs-tools#742` preserves the original Jul Lang / Codex commit stack instead of carrying the port as one squashed commit; tools-only adaptations are separate Komzpa commits.
- The latest tail commits fix shrink restart handling in discard retry and cached-tail invalidation paths. Earlier tail commits also let online smaller-size `bcachefs device resize` requests reach the resize worker and resize the tools-only `buckets_nouse` bitmap with the bucket arrays.
- It remains intentionally draft/WIP, because this is still experimental shrink code and should not be treated as safe for important filesystems yet.
- Current `koverstreet/bcachefs-tools#742` head `d432f6894649123cc1e6fde2157b967f2662094e` is merge-clean with 17/17 visible GitHub CI checks passing. Local validation also passed `git diff --check`, `BINDGEN=/home/kom/.cargo/bin/bindgen make -j32 bcachefs`, `./target/release/bcachefs device resize --help`, `make -C /lib/modules/$(uname -r)/build M=$PWD/fs -j32 modules`, and the full local shrink suite.
- Companion draft ktest coverage is `koverstreet/ktest#85` at head `0bdec5d426122b740faefbda88c5d2fcbfe1f7e5`, merge-clean with GitHub CI green. It also preserves the imported Jul Lang test commit stack, leaves out offline shrink with a separate online-scope adaptation commit because `koverstreet/bcachefs-tools#742` still leaves offline shrink rejected, and adds a separate retarget-race fix.

So for this issue cluster, `koverstreet/bcachefs-tools#742` is the current bcachefs-tools-side implementation artifact, `koverstreet/ktest#85` is the current ktest proof artifact, and `koverstreet/bcachefs#1073` remains the original/provenance implementation surface.

**Comment by jullanggit** [comment]
@Komzpa it seems like your PR references don't target the right repositories

**Comment by koverstreet** [comment]
Closing as a duplicate — filesystem shrink is tracked in the active bcachefs-tools tracker at koverstreet/bcachefs-tools#268, with implementation work in flight (bcachefs-tools#741 / #742). Please follow that issue.

**Comment by Malix-Labs** [comment]
koverstreet/bcachefs-tools#741

**Comment by jullanggit** [comment]
The original pull request / draft for the actual implementation would be at https://github.com/koverstreet/bcachefs/pull/1073
